betterproto/__init__.py

```diff
@@ -426,6 +430,81 @@ def parse_fields(value: bytes) -> Generator[ParsedField, None, None]:
 T = TypeVar("T", bound="Message")
 
 
+class ProtoClassMetadata:
+    cls: "Message"
+
+    def __init__(self, cls: "Message"):
+        self.cls = cls
+        by_field = {}
+        by_group = {}
+
+        for field in dataclasses.fields(cls):
+            meta = FieldMetadata.get(field)
+
+            if meta.group:
+                # This is part of a one-of group.
+                by_field[field.name] = meta.group
+
+                by_group.setdefault(meta.group, set()).add(field)
+
+        self.oneof_group_by_field = by_field
+        self.oneof_field_by_group = by_group
+
+    def init_default_gen(self):
+        default_gen = {}
+
+        for field in dataclasses.fields(self.cls):
+            meta = FieldMetadata.get(field)
+            default_gen[field.name] = self.cls._get_field_default_gen(field, meta)
+
+        self.default_gen = default_gen
+
+    def init_cls_by_field(self):
+        field_cls = {}
+
+        for field in dataclasses.fields(self.cls):
+            meta = FieldMetadata.get(field)
+            if meta.proto_type == TYPE_MAP:
+                assert meta.map_types
+                kt = self.cls._cls_for(field, index=0)
+                vt = self.cls._cls_for(field, index=1)
+                Entry = dataclasses.make_dataclass(
+                    "Entry",
+                    [
+                        ("key", kt, dataclass_field(1, meta.map_types[0])),
+                        ("value", vt, dataclass_field(2, meta.map_types[1])),
+                    ],
+                    bases=(Message,),
+                )
+                make_protoclass(Entry)
+                field_cls[field.name] = Entry
+                field_cls[field.name + ".value"] = vt
+            else:
+                field_cls[field.name] = self.cls._cls_for(field)
+
+        self.cls_by_field = field_cls
+
+    def __getattr__(self, item):
+        # Lazy init because forward reference classes may not be available at the beginning.
+        if item == 'default_gen':
+            self.init_default_gen()
+            return self.default_gen
+
+        if item == 'cls_by_field':
+            self.init_cls_by_field()
+            return self.cls_by_field
+
+
+def make_protoclass(cls):
+    setattr(cls, "_betterproto", ProtoClassMetadata(cls))
+
+
+def protoclass(*args, **kwargs):
```

Comment:
Am I right in my understanding, that this causes users to have to use the created `protoclass` instead of creating dataclasses?

However:

1. It's auto-generated
2. It behaves like a dataclass

So this doesn't seem to break anything.

Reply:
Unless I'm missing something this looks like classes the inherit from `betterproto.Message` with a regular dataclass decorator will break if the user upgrades due to _betterproto not being set.

Breaking changes may be OK if this comes as a major version bump, but would still be better to avoid it or at least try and handle it gracefully.

I suggest accessing _betterproto via a property that checks whether it is set, and degrades gracefully for dataclasses lacking the extra annotation, maybe with a warning message. If going this route then there should be a test for it.

Reply:
`protoclass` is a wrapper of `dataclasses.dataclass`. When this decorator is applied, the `dataclass` is still applied and the data class is still generated. The extra thing `protoclass` does, is to call `make_protoclass` to scan metadata of fields, and store in attribute `_betterproto` of the generated class.

In this way, the scan operation is done once per class, instead of per object, or per method call.

So this change requires to replace `@dataclasses.dataclass` with `@betterproto.protoclass` in generated python files. Which means users need to rerun `protoc` to update.

Reply:
@jameslan Thanks for clarifying.

Reply:
Now the extra decorator is removed. The metadata is populated at the first access, instead of when data class is created.